cli/src/main/kotlin/tools/samt/cli/DiagnosticFormatter.kt

```diff
@@ -0,0 +1,505 @@
+package tools.samt.cli
+
+import com.github.ajalt.mordant.rendering.TextColors
+import com.github.ajalt.mordant.rendering.TextColors.*
+import com.github.ajalt.mordant.rendering.TextStyles.*
+import com.github.ajalt.mordant.terminal.Terminal
+import tools.samt.common.*
+
+class DiagnosticFormatter(
+    private val diagnosticController: DiagnosticController,
+    private val terminalWidth: Int,
+) {
+    companion object {
+        private const val CONTEXT_ROW_COUNT = 3
+
+        // FIXME: this is a bit of a hack to get the terminal width
+        //        it also means we're assuming this output will only ever be printed in a terminal
+        //        i don't actually know what happens if it doesn't run in a tty setting
+        fun format(controller: DiagnosticController, terminalWidth: Int = Terminal().info.width): String {
+            val formatter = DiagnosticFormatter(controller, terminalWidth)
+            return formatter.format()
+        }
+
+        private fun rowContextLowerBound(row: Int): Int = maxOf(0, row - CONTEXT_ROW_COUNT)
+        private fun rowContextUpperBound(row: Int, sourceFile: SourceFile): Int = minOf(sourceFile.sourceLines.lastIndex, row + CONTEXT_ROW_COUNT)
+    }
+
+    private fun format(): String = buildString {
+
+        // print context-less messages
+        val globalMessages = diagnosticController.globalMessages
+        globalMessages.forEachIndexed { index, message ->
+            appendLine("⎯".repeat(terminalWidth))
```

Comment:
This dubious character doesn't render on Windows, I recommend we switch it here **and** below to the equivalent [Box-drawing character](https://en.wikipedia.org/wiki/Box-drawing_character) `─`
```suggestion
            appendLine("─".repeat(terminalWidth))
```